// server/twitch/api/AbstractAPI.js
"use strict";
var request = require('request');
var qs = require('querystring');
var fileLog_1 = require('../../lib/fileLog');
var messages_1 = require('../../common/messages');
var twitch_config_1 = require('../twitch-config');
var tsw_config_1 = require('../../tsw/tsw-config');
var req = request;
if (tsw_config_1.HTTP_PROXY) {
    req = request.defaults({ 'proxy': tsw_config_1.HTTP_PROXY });
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = AbstractAPI;
var AbstractAPI = (function () {
    function AbstractAPI() {
        this.method = 'GET';
        this.oauth = false;
    }
    AbstractAPI.prototype.request = function (query, cb) {
        var me = this;
        var logger = plug('logger');
        query.client_id = twitch_config_1.OAUTH.CLIENT_ID;
        query.state = Math.random();
        var apiURL = twitch_config_1.BASE_URI + this.pathname;
        var apiHeaders = {
            'Client-ID': twitch_config_1.OAUTH.CLIENT_ID
        };
        var apiPostData;
        if (this.method === 'GET') {
            apiURL += '?' + qs.stringify(query);
        }
        else if (this.method === 'POST') {
            apiPostData = query;
        }
        if (this.oauth) {
            apiHeaders['Authorization'] = 'OAuth ' + query.accessToken;
        }
        logger.info("[TWITCH API]\n                     url=" + apiURL + "\n                     headers=" + (apiHeaders ? JSON.stringify(apiHeaders) : ''));
        request({
            method: me.method,
            url: apiURL,
            proxy: tsw_config_1.HTTP_PROXY,
            headers: apiHeaders,
            form: apiPostData
        }, function (err, res, body) {
            var json;
            if (!err && res.statusCode == 200) {
                try {
                    json = JSON.parse(body);
                }
                catch (e) {
                    logger.error(me.pathname + " \u89E3\u6790Twitch\u8FD4\u56DE\u503C\u5931\u8D25 " + e.message);
                    err = messages_1.default.ERR_PARSE_JSON_FAILED;
                }
            }
            if (json) {
                logger.info('请求Twitch接口成功');
                fileLog_1.default.info(JSON.stringify(json));
                cb(null, json);
            }
            else {
                logger.error("err=" + (err || '') + ", statusCode=" + (res ? res.statusCode : '') + ", body=" + JSON.stringify(body));
                cb(messages_1.default.ERR_REQUEST_TWITCH_API_FAILED, {});
            }
        });
    };
    return AbstractAPI;
}());
var log2file = function (s) {
    "use strict";
};
