// server/common/view/pug-render.js
const pug = require('pug');
const messages = require('../../common/messages');
const debug = process.env.NODE_ENV === 'development';

module.exports = {

    /**
     * 渲染pug文件
     * @param {String} pugFile 模版文件路径，不允许为空且必须唯一
     * @param {Object} data 模版数据
     * @param {Function} cb
     */
    render: (pugFile, data)=> {
        "use strict";

        let logger = plug('logger');

        let req = window.request;
        let res = window.response;

        pug.renderFile(pugFile, {
            //basedir: require('path').join(__dirname, '../../'),
            cache: true,
            pretty: ' ',
            debug: debug,
            compileDebug: debug,
            data: Object.assign({}, {req: req}, data)
        }, (err, html)=> {
            if (err) {
                logger.error(err);
                err = messages.ERR_PUG_RENDER_ERROR;
                res.write(JSON.stringify(err));
            } else {
                res.writeHead(200, {'Content-Type': 'text/html'});
                res.write(html);
            }
            res.end();
        });
    }
};
