// server/lib/log.ts
import * as path from 'path'
import * as winston from 'winston'
import * as WinstonDailyRotateFile from 'winston-daily-rotate-file'
import ConsoleTransportInstance = winston.ConsoleTransportInstance;
import GenericTextTransportOptions = winston.GenericTextTransportOptions;

const PATH_LOG_FILE = path.join(__dirname, '../../../log/run.log');

const transports = [];

const formatter = (o) => {
    let pFile = module.parent.filename;
    let pName = pFile.split(/[\/\\]/).slice(-2).join('/').replace(/\.js$/, '');
    let time = o.timestamp();
    return `[${time.getFullYear()}-${time.getMonth() + 1}-${time.getDate()} ${time.getHours()}:${time.getMinutes()}:${time.getSeconds()}] [${o.level.substr(0, 1).toUpperCase()}] [${pName}] ${o.message || ''}`;
};

if (process.env.NODE_ENV === 'test' ||
    process.env.NODE_ENV === 'production') {

    transports.push(new WinstonDailyRotateFile({
        filename: PATH_LOG_FILE,
        timestamp: function () {
            return new Date();
        },
        datePattern: 'yyyy-MM-dd.',
        prepend: true,
        level: 'info'
    }));
}
else {
    transports.push(new (winston.transports.Console)({
        level: 'debug',
        timestamp: function () {
            return new Date();
        },
        formatter: formatter
    }));
}

export default new (winston.Logger)({
    transports: transports,
    formatter: formatter
});
