// server/lib/log.js
"use strict";
var path = require('path');
var winston = require('winston');
var WinstonDailyRotateFile = require('winston-daily-rotate-file');
var PATH_LOG_FILE = path.join(__dirname, '../../../log/run.log');
var transports = [];
var formatter = function (o) {
    var pFile = module.parent.filename;
    var pName = pFile.split(/[\/\\]/).slice(-2).join('/').replace(/\.js$/, '');
    var time = o.timestamp();
    return "[" + time.getFullYear() + "-" + (time.getMonth() + 1) + "-" + time.getDate() + " " + time.getHours() + ":" + time.getMinutes() + ":" + time.getSeconds() + "] [" + o.level.substr(0, 1).toUpperCase() + "] [" + pName + "] " + (o.message || '');
};
if (process.env.NODE_ENV === 'test' ||
    process.env.NODE_ENV === 'production') {
    transports.push(new WinstonDailyRotateFile({
        filename: PATH_LOG_FILE,
        timestamp: function () {
            return new Date();
        },
        datePattern: 'yyyy-MM-dd.',
        prepend: true,
        level: 'info'
    }));
}
else {
    transports.push(new (winston.transports.Console)({
        level: 'debug',
        timestamp: function () {
            return new Date();
        },
        formatter: formatter
    }));
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = new (winston.Logger)({
    transports: transports,
    formatter: formatter
});
